sympy/matrices/matrices.py

```diff
@@ -1547,6 +1547,13 @@ def _matrix_pow_by_jordan_blocks(self, num):
         def jordan_cell_power(jc, n):
             N = jc.shape[0]
             l = jc[0, 0]
+            if (num < 0) == True and l == 0:
+                raise ValueError("Matrix det == 0; not invertible")
+            if (num < 2) == True:
```

Comment:
I suggest that only non-negative integer exponents are allowed if there are nilpotent Jordan blocks (of size > 1). That could probably include floats that are equal to integers.